cmake/app/boilerplate.cmake

```diff
@@ -306,6 +306,33 @@ foreach(root ${BOARD_ROOT})
       else()
         list(APPEND NOT_FOUND_SHIELD_LIST ${s})
       endif()
+
+      # search for shield/boards/board.overlay file
+      if(EXISTS ${shield_dir}/${s}/boards/${BOARD}.overlay)
+        # add shield/board overlay to the shield overlays list
+        list(APPEND
+          shield_dts_files
+          ${shield_dir}/${s}/boards/${BOARD}.overlay
+        )
+      endif()
+
+      # search for shield/shield.conf file
+      if(EXISTS ${shield_dir}/${s}/${s}.conf)
```

Comment:
I think shields should follow the same rule as boards, ie configure the HW, but not activate features, which is application responsibility. So in general, having Kconfig.defconfig (conditional confguration), rather than .conf or def_config that set Kconfig symbols.

Reply:
I don't agree entirely.  The user has explicitly added the SHIELD configuration to the build.  In my head this is exactly like specifying overlay-*.conf (with the added benefit that SHIELD configuration is accessible across all samples).

The Kconfig.defconfig file is processed **every time** the board is built for regardless of the configuration.  Hence the conditional nature.  This is not the same IMHO.

I *do* agree WNC-M14A2A .conf files are a bit messy:
1) Settings like this are fine:
```
# Enable WNC-M14A2A LTE-M Modem
CONFIG_MODEM=y
CONFIG_MODEM_SHELL=y
CONFIG_MODEM_WNCM14A2A=y
CONFIG_UART_INTERRUPT_DRIVEN=y
```
2) The following are definitely hackish.  They are specific to the transport medium and driver, not necessarily the shield:
```
# extend retry timing to 20 seconds for LTE/LTE-M
CONFIG_COAP_INIT_ACK_TIMEOUT_MS=20000

# Hack: disable IPv6 for now
CONFIG_NET_IPV6=n
CONFIG_NET_CONFIG_NEED_IPV6=n
```

I'm looking to fix these in a different way in the future.

3) Most SHIELD setups won't need the boards/* additions like `frdm_k64f.conf` which sets `CONFIG_UART_MCUX_2=y`.  This is due to the serial port being on a non-standard Arduino header pin (standard serial is `CONFIG_UART_MCUX_3`).
Most of the settings in `nrf52840_pca10056.conf` can probably be abstracted like @MaureenHelm mentioned where the board sets up the following if `CONFIG_ARDUINO_SERIAL` is enabled:
```
CONFIG_GPIO_NRF_P1=y
CONFIG_UART_1_NRF_UARTE=y
```

Not sure what to do about the: `CONFIG_UART_1_NRF_FLOW_CONTROL=y` setting.  This particular modem (driver) needs it and it's not a setting controlled by DTS.  This one might need to stay in `boards/nrf52840_pca10056.conf`.

Reply:
> I don't agree entirely. The user has explicitly added the SHIELD configuration to the build. In my head this is exactly like specifying overlay-*.conf (with the added benefit that SHIELD configuration is accessible across all samples).

OTOH, we can have boards with the same HW component (be it a modem, a sensor, ..) out of the box. From application point of view, to use this HW component, being direct part of the BSP or added as a shield, same configuration should be done (MODEM, INTERRUPT_DRIVEN, ..).
So shield configuration should add the missing configuration bits between application and the board default one.

> The Kconfig.defconfig file is processed every time the board is built for regardless of the configuration. Hence the conditional nature. This is not the same IMHO.

I've missed your point here.

Reply:
@erwango 

> I've missed your point here.

I was trying to distinguish between a file that would only be processed if the user includes it on the build line vs. Kconfig.defconfig which is always processed.  Can ignore see below:

I added 1 more commit as an example for enabling the SARA-R4 modem on the Particle.io Boron board here: https://github.com/zephyrproject-rtos/zephyr/pull/14057/commits/40321361e59c6364fde089b914221b14258a51fe
(Boron has built-in modem)

This is a concrete example we can discuss as it uses the more traditional method of Kconfig.defconfig: if CONFIG_MODEM is enabled, it chooses the SARA_R4 driver.  The obvious problem with the Boron board is that it has several methods of communication:
- BLE 6lowpan (via CONFIG_NET_L2_BT)
- 802.15.4 via (via CONFIG_NET_L2_IEEE802154 or CONFIG_NET_L2_OPENTHREAD)
- SARA-R4 modem (via CONFIG_MODEM)

The only way Zephyr currently has of dealing with these is to add overlay-*.conf files into each sample:
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-bt.conf
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-802154.conf
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-ot.conf

Such overlay files are typically carbon copies of each other across all of the samples and they aren't scaling well.  Consider these 2 overlay-bt.conf files:
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/lwm2m_client/overlay-bt.conf
and
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-bt.conf

In a perfect world we have a `boards/common-overlays` dir and this is part of the overlay-*.conf search path when checking -DOVERLAY_CONFIG= files.   This is something we could do regardless of shield-based changes.

Let's also take a look at the current contents of:
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/lwm2m_client/overlay-wncm14a2a.conf

This is incredibly similar to the wnc_m14a2a.conf file added under boards/shields/wnc_m14a2a/.

Let's examine your proposal of doing the settings as conditional in a Kconfig.defconfig file:
- Move not only the shield's .conf file into Kconfig.defconfig format, but also create conditional handling for all of the supported boards as well (if there are boards/*.conf) because there are dependencies in those files which might or might not be enabled.
- After that's done, the user would still need to add -DOVERLAY_CONFIG=* with a generic opt-in overlay file which enables the main piece of hardware / functionality.

This doesn't *seem* like a better solution unless I'm missing an example where the user would specifically add -DSHIELD=* to his build line, but **NOT** want the functionality of the shield enabled.   Perhaps you have a good example in mind?